File: Python/calc1.py
import os


# Color definitions
END = "\033[0m"
GREEN = "\033[0;32m"
BLUE = "\033[0;94m"
REDBOLD = "\033[1;31m"
ORANGEBOLD = "\033[1;33m"
WHITEUNDER = "\033[4;37m"


# Functions used by calculator
def add(a: float, b: float) -> float:
    return a + b

def sub(a: float, b: float) -> float:
    return a - b

def mul(a: float, b: float) -> float:
    return a * b

def div(a: float, b: float) -> float:
    return a / b

def exp(a: float, b: float) -> float:
    # Not figured out yet how to handle roots with help of iterations
    if b != int(b):
        return a ** b
    # 0^0 is not defined
    if a ==  b == 0:
        raise(ZeroDivisionError)

    # Check if b is negative and take absolute value of b
    positiveExp = b > -1
    b = b if positiveExp else -b

    # Calculate exponent manually by iterating product
    product = 1
    for _ in range(int(b)):
        product *= a
    
    return product if positiveExp else 1 / product

# Display a small message and exit program
def exitAppl() -> None:
    print("\nThank you for using the calculator")
    print("Exiting program...")
    exit()


# Main function of the program
def main():
    # Tuple of tuples containting (signe, name, function)
    modes = (
        (f"{BLUE}+{END}", f"{BLUE}Addition{END}", add),
        (f"{BLUE}-{END}", f"{BLUE}Substraction{END}", sub),
        (f"{BLUE}*{END}", f"{BLUE}Multiplicaton{END}", mul),
        (f"{BLUE}/{END}", f"{BLUE}Division{END}", div),
        (f"{BLUE}^{END}", f"{BLUE}Exponentation{END}", exp)
    )

    while True:
        # Clear terminal
        os.system("clear" if os.name != "nt" else "cls")

        # Print start message
        print(f"{WHITEUNDER}Python CLI calc V1.0{END}\n")
        print("Choose a function")
        print(f"{GREEN}1{END} - {BLUE}Add        {GREEN}3{END} - {BLUE}Multiply   {GREEN}5{END} - {BLUE}Exponentiation{END}")
        print(f"{GREEN}2{END} - {BLUE}Subtract   {GREEN}4{END} - {BLUE}Divide     {GREEN}6{END} - {BLUE}Quit{END}")

        try:
            # Get user input
            a = int(input(">>> ")) - 1

            # Check if user wants to quit
            if a == len(modes) or a == -len(modes)-1:
                exitAppl()

            mode = modes[a]

            # Display information of function and get all inputs  
            print(f"\n{BLUE}{mode[1]} - Please choose two numbers a and b to calculate {GREEN}a{mode[0]}{GREEN}b{END}")
            a = float(input("a: "))
            b = float(input("b: "))
            # Print result
            print(f"{BLUE}Result: {GREEN}{mode[2](a, b)}{END}")
            
            # Get enter keypress to continue the program
            input("Press enter to restart...")

        # Handle exceptions and restart program
        except ValueError:
            # User did not enter a number
            print(f"{REDBOLD}ERROR:{END} Please only enter numbers")
            input("Press enter to restart...")
        except IndexError:
            # User chose a function which does not exist
            print(f"{REDBOLD}ERROR:{END} Please enter a number between {-len(modes)} and {len(modes)+1}")
            input("Press enter to restart...")
        except ZeroDivisionError:
            # User wanted to divide by zero
            print(f"{ORANGEBOLD}WARNING:{END} Please don't divide by 0")
            input("Press enter to restart...")
        except KeyboardInterrupt:
            # Quit program properly on ctrl + c
            exitAppl()
        except Exception:
            # Make sure program does not error out
            print(f"{REDBOLD}ERROR:{END} Something went wrong")
            input("Press enter to restart...")


# Run file if file is directly executed
if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        exit()

File: C/calc1.c
#include <stdio.h>
#include <stdlib.h>

// Macros to define colors and get size of an array 
#define len(a) sizeof(a)/sizeof(*a)
#define END "\e[0m"
#define PURPLE "\e[0;35m"
#define CYAN "\e[0;36m"
#define WHITEUNDER "\e[4;37m"
#define ERROR "\e[1;37m[\e[1;31mERROR\e[1;37m]\e[0m"
#define WARNING "\e[1;37m[\e[1;33mWARNING\e[1;37m]\e[0m"

// Gets turned to 1 if zero division occurs 
unsigned char zeroDivision = 0;


// Fucntions used by the calculator
double absolute(double a)
{
    return (a > 0)? a : -a; 
}

double add(double a, double b)
{
    return (a + b);
}

double subtract(double a, double b)
{
    return (a - b);
}

double multiply(double a, double b)
{
    return (a * b);
}

double divide(double a, double b)
{
    // Check if user does not divide by 0
    if (absolute(b) > 0) {
        return (a / b);
    }
    zeroDivision = 1;
    return 0;
}

double power(double a, double b)
{
    // Check if b is negative and takeabsolute value of b
    unsigned char negativPow = b < 0;
    b = absolute(b);

    // 0^0 is not defined
    if (a == 0 && b == 0) {
        zeroDivision = 1;
        return 0;
    }

    // Calculate exponent manually by iterating product
    double product = 1;
    for (int _ = 0; _ < b; _++) {
        product *= a;
    }

    return (!negativPow)? product : divide(1, product);
}

// Display restart message, clear stdin and wait until enter keypress
static void wait_for_restart()
{
    char ch;
    printf("\nPress enter to restart...\n");
    while ((ch=getchar()) != '\n' && ch != EOF);
    fgetc(stdin);
}


int main()
{
    // input to get mode and notError to check for errors
    unsigned char input, notError;
    // Doubles to store calculator inputs and output
    double a, b, result;
    // Array containing `{signe, name}`
    const char *modes[][2]= {
        {PURPLE"+"END, PURPLE"Addition"END},
        {PURPLE"-"END, PURPLE"Subtraction"END},
        {PURPLE"*"END, PURPLE"Multiplication"END},
        {PURPLE"/"END, PURPLE"Division"END},
        {PURPLE"^"END, PURPLE"Exponentation"END}
    };
    // Array containing function pointers
    double (*functions[])(double, double) = {
        add, subtract, multiply, divide, power
    };


    while (1) {
        // Clear terminal
        system("clear");
        // Set zero division marker to false
        zeroDivision = 0;

        // Print start message
        printf(WHITEUNDER"C CLI calc V1.0"END"\n\n");
        printf("Choose a function\n");
        printf(CYAN"1"END" - "PURPLE"Add         "CYAN"3"END" - "PURPLE"Multiply    "CYAN"5"END" - "PURPLE"Exponentation"END"\n");
        printf(CYAN"2"END" - "PURPLE"Subtract    "CYAN"4"END" - "PURPLE"Divide      "CYAN"6"END" - "PURPLE"Quit"END"\n");
        printf(">>> ");

        // Get input and decrease it
        notError = scanf("%hhu", &input);
        input--;

        // Mode selection error handling
        if (!notError) {
            // User did not enter numbers
            printf("\n"ERROR" Please only enter whole numbers!\n");
            wait_for_restart();
            continue;
        } else if (input < -1 || input > len(functions)) {
            // User chose a function which does not exist
            printf("\n"ERROR" Please enter a number between 1 and %li\n", len(functions)+1);
            wait_for_restart();
            continue;
        } else if (input == len(functions)) {
            // User want to exit the program
            printf("Exiting program...\n");
            return 0;
        }

        // Display information of function and get all inputs
        printf("You chose: %s\n", modes[input][1]);
        printf("Please enter two numbers a and b to calculate "CYAN"a%s"CYAN"b"END"\n", modes[input][0]);

        printf("a: ");
        notError = scanf("%lf", &a);
        if (notError) {
            printf("b: ");
            notError = scanf("%lf", &b);
        }
        
        // Calculator input error handling
        if (!notError) {
            // User did not enter numbers
            printf("\n"ERROR" Please only enter numbers!\n");
            wait_for_restart();
            continue;
        }

        // Call calculator and check if user did not divide by zero
        result = functions[input](a, b);
        if (zeroDivision) {
            // User wanted to divide by zero
            printf("\n"WARNING" Please do not divide by 0!\n");
            wait_for_restart();
            continue;
        }

        // Print result
        printf(PURPLE"Result:"CYAN" %lf\n"END, result);
        wait_for_restart();
    }

    return 0;
}